js/util/classify_rings.js

```diff
@@ -35,10 +36,9 @@ function classifyRings(rings, maxRings) {
     if (maxRings > 1) {
         len = polygons.length;
         for (var j = 0; j < len; j++) {
-            polygon = polygons[j];
-            truncated = [polygon.shift()];
-            polygon.sort(sortByArea);
-            polygons[j] = truncated.concat(polygon.slice(0, maxRings - 1));
+            if (polygons[j].length <= maxRings) continue;
+            quickselect(polygons[j], maxRings - 1, 1, polygon.length - 1, sortByArea);
+            polygons[j] = polygon.slice(0, maxRings);
```

Comment:
It seems that the 2nd arg to `quickselect` should be just `maxRings` — `k` is position independent of `left`.